python/kvikio/nvcomp.py

```diff
@@ -79,147 +107,240 @@ def compress(self, data: cp.ndarray) -> cp.ndarray:
+                size of the buffer that must be allocated before calling compress.
+            "num_chunks": The number of configured chunks to compress the data over
+        }
+        """
+        return self._manager.configure_compression(data_size)
+
+    def configure_decompression_with_compressed_buffer(
+        self, data: cp.ndarray
+    ) -> cp.ndarray:
+        """Return the decompression configuration object.
 
         Parameters
         ----------
-        dtype: cp.dtype
-            The input buffer dtype this LZ4Compressor will compress.
+        data: cp.ndarray
+            A GPU buffer of previously compressed data.
+
+        Returns
+        -------
+        dict {
+            "decomp_data_size": The size of each decompression chunk.
+            "num_chunks": The number of chunks that the decompressed data is returned
+            in.
+        }
         """
-        self.dtype = cp.dtype(dtype)
-        self.compressor = _lib._LZ4Compressor()
-        self.decompressor = _lib._LZ4Decompressor()
-        self.s = cp.cuda.Stream()
+        return self._manager.configure_decompression_with_compressed_buffer(data)
 
-    def compress(self, data: cp.ndarray) -> cp.ndarray:
-        """Compress a buffer.
+    def get_required_scratch_buffer_size(self) -> int:
+        """Return the size of the optional scratch buffer.
 
         Returns
         -------
-        cp.ndarray
-            A GPU buffer of compressed bytes.
+        int
         """
-        # TODO: An option: check if incoming data size matches the size of the
-        # last incoming data, and reuse temp and out buffer if so.
-        data_size = data.size * data.itemsize
-        self.compress_temp_size = np.zeros((1,), dtype=np.int64)
-        self.compress_out_size = np.zeros((1,), dtype=np.int64)
-        self.compressor.configure(
-            data_size, self.compress_temp_size, self.compress_out_size
-        )
+        return self._manager.get_required_scratch_buffer_size()
 
-        self.compress_temp_buffer = cp.zeros(
-            (self.compress_temp_size[0],), dtype=cp.uint8
-        )
-        self.compress_out_buffer = cp.zeros(
-            (self.compress_out_size[0],), dtype=cp.uint8
-        )
-        # Weird issue with LZ4 Compressor - if you pass it a gpu-side out_size
-        # pointer it will error. If you pass it a host-side out_size pointer it will
-        # segfault.
-        self.gpu_out_size = cp.array(self.compress_out_size, dtype=np.int64)
-        self.compressor.compress_async(
-            data,
-            data_size,
-            self.compress_temp_buffer,
-            self.compress_temp_size,
-            self.compress_out_buffer,
-            self.gpu_out_size,
-            self.s.ptr,
-        )
-        return self.compress_out_buffer[: self.compress_out_size[0]]
+    def set_scratch_buffer(self, new_scratch_buffer: cp.ndarray) -> None:
```

Comment:
Are you asking if the underlying C++ has a danger of freeing the scratch buffer that we're specifying here?

Or are you asking if the underlying library is going to take a piece of ownership of the scratch buffer and prevent it from being freed?

Reply:
I was just wondering if users should be able to unset the scratch buffer and use it for something else (or free it) without having to delete the `nvCompManager`.

Reply:
I can't make promises about what nvcomp does under the hood, but nowhere in the bindings do we take ownership of the scratch buffer. I assume that they don't either, since they're just taking a pointer as always.

Reply:
This is a persistent state that you're providing the NvcompManager. Basically nvcompManager will use this buffer for all future compress/decompress operations until you specify a new pointer. You're free to use it for something else in the same CUDA stream (it doesn't store state between kernels). If you free it, you'll synchronize the stream that nvcompManager's been using. Then you need to provide a new scratch buffer before before the next call. Currently there's no way to "clear" this variable (which would lead to nvcompManager allocating internally if necessary).

Reply:
Thanks for clarifying @eschmidt-nvidia